src/libs/kata-sys-util/src/hooks.rs

```diff
@@ -0,0 +1,519 @@
+// Copyright (c) 2019-2021 Alibaba Cloud
+// Copyright (c) 2019-2021 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::collections::HashMap;
+use std::ffi::OsString;
+use std::hash::{Hash, Hasher};
+use std::io::{self, Read, Result};
+use std::path::Path;
+use std::time::Duration;
+
+use subprocess::{ExitStatus, Popen, PopenConfig, PopenError, Redirection};
+
+use crate::{eother, sl};
+
+const DEFAULT_HOOK_TIMEOUT_SEC: i32 = 10;
+
+#[derive(Debug, Default, Clone)]
+struct HookKey(oci::Hook);
+
+impl From<&oci::Hook> for HookKey {
+    fn from(hook: &oci::Hook) -> Self {
+        HookKey(hook.clone())
+    }
+}
+
+impl PartialEq for HookKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.0 == other.0
+    }
+}
+
+impl Eq for HookKey {}
+
+impl Hash for HookKey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.path.hash(state);
+        self.0.args.hash(state);
+        self.0.env.hash(state);
+        self.0.timeout.hash(state);
+    }
+}
+
+/// Execution state of OCI hooks.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum HookState {
+    /// Hook is pending for executing/retry.
+    Pending,
```

Comment:
Is there any advantage in adding a `Retry(u64)` here to make the state distinct and to make it easy to track the number of attempted retries?